### PR TITLE
Add a specialized instruction for `.nil?` calls

### PR DESCRIPTION
**Pull request by tenderlove**
This commit adds a specialized instruction for called to `.nil?`.  It is
about 27% faster than master in the case where the object is nil or not
nil.  In the case where an object implements `nil?`, I think it may be
slightly slower.  Here is a benchmark:

```ruby
require "benchmark/ips"

class Niller
  def nil?; true; end
end

not_nil = Object.new
xnil = nil
niller = Niller.new

Benchmark.ips do |x|
  x.report("nil?")    { xnil.nil? }
  x.report("not nil") { not_nil.nil? }
  x.report("niller")   { niller.nil? }
end
```

On Ruby master:

```
[aaron@TC ~/g/ruby (master)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   429.195k i/100ms
             not nil   437.889k i/100ms
              niller   437.935k i/100ms
Calculating -------------------------------------
                nil?     20.166M (± 8.1%) i/s -    100.002M in   5.002794s
             not nil     20.046M (± 7.6%) i/s -     99.839M in   5.020086s
              niller     22.467M (± 6.1%) i/s -    112.111M in   5.013817s
[aaron@TC ~/g/ruby (master)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   449.660k i/100ms
             not nil   433.836k i/100ms
              niller   443.073k i/100ms
Calculating -------------------------------------
                nil?     19.997M (± 8.8%) i/s -     99.375M in   5.020458s
             not nil     20.529M (± 7.0%) i/s -    102.385M in   5.020689s
              niller     21.796M (± 8.0%) i/s -    108.110M in   5.002300s
[aaron@TC ~/g/ruby (master)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   402.119k i/100ms
             not nil   438.968k i/100ms
              niller   398.226k i/100ms
Calculating -------------------------------------
                nil?     20.050M (±12.2%) i/s -     98.519M in   5.008817s
             not nil     20.614M (± 8.0%) i/s -    102.280M in   5.004531s
              niller     22.223M (± 8.8%) i/s -    110.309M in   5.013106s

```

On this branch:

```
[aaron@TC ~/g/ruby (specialized-nilp)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   468.371k i/100ms
             not nil   456.517k i/100ms
              niller   454.981k i/100ms
Calculating -------------------------------------
                nil?     27.849M (± 7.8%) i/s -    138.169M in   5.001730s
             not nil     26.417M (± 8.7%) i/s -    131.020M in   5.011674s
              niller     21.561M (± 7.5%) i/s -    107.376M in   5.018113s
[aaron@TC ~/g/ruby (specialized-nilp)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   477.259k i/100ms
             not nil   428.712k i/100ms
              niller   446.109k i/100ms
Calculating -------------------------------------
                nil?     28.071M (± 7.3%) i/s -    139.837M in   5.016590s
             not nil     25.789M (±12.9%) i/s -    126.470M in   5.011144s
              niller     20.002M (±12.2%) i/s -     98.144M in   5.001737s
[aaron@TC ~/g/ruby (specialized-nilp)]$ ./ruby compil.rb
Warming up --------------------------------------
                nil?   467.676k i/100ms
             not nil   445.791k i/100ms
              niller   415.024k i/100ms
Calculating -------------------------------------
                nil?     26.907M (± 8.0%) i/s -    133.755M in   5.013915s
             not nil     25.319M (± 7.9%) i/s -    125.713M in   5.007758s
              niller     19.569M (±11.8%) i/s -     96.286M in   5.008533s
```

Co-Authored-By: Ashe Connor <kivikakk@github.com>

### CODE REVIEW

**Comment by k0kubun** [comment]
The micro benchmark looks good. I'm also interested in a non-trivial benchmark. I guess at least something encouraged you to optimize `nil?`.

I benchmarked this with a Rails scaffold (bin/bench of https://github.com/k0kubun/railsbench) and Noah's RSB (./runners/current_ruby.rb of https://github.com/noahgibbs/rsb), but these benchmarks did not generate a significant difference on my machine https://gist.github.com/k0kubun/15097ca0c919a9bcbdb56714cb44dd67. They may be overriding it or not going through `nil?`. (Optcarrot seems not using it either.)

File: object.c [hunk 1]
```diff
@@ -1664,7 +1664,7 @@ rb_true(VALUE obj)
  */
 
 
-static VALUE
+VALUE
```
Comment (k0kubun) [inline object.c]:
```suggestion
MJIT_FUNC_EXPORTED VALUE
```
To fix test failures, this function needs to be exported.

**Comment by ioquatix** [comment]
What is faster? `unless x` vs `if x == nil` or `if x.nil?`? What is most common?

**Comment by tenderlove** [comment]
> I guess at least something encouraged you to optimize `nil?`.

I saw we have many calls for `.nil?` in our application (I noted that [here](https://bugs.ruby-lang.org/issues/15902)).  We have similar number of calls to `.nil?` as `.length` and `.empty?`, and both of those have optimized instructions [1](https://github.com/ruby/ruby/blob/90f59dbb8121131de93846697f8bf78565a04b1c/insns.def#L1355-L1367) and [2](https://github.com/ruby/ruby/blob/90f59dbb8121131de93846697f8bf78565a04b1c/insns.def#L1384-L1395).

As I said in the [redmine ticket](https://bugs.ruby-lang.org/issues/15902), I'm not sure how hot those call sites are.  But at least in our application `.nil?` checks are pretty common.

**Comment by tenderlove** [comment]
> What is faster? `unless x` vs `if x == nil` or `if x.nil?`? What is most common?

I'll guess that the `unless` and `if x == nil` forms are faster.  It's hard for me to grep for the unless forms, but `== nil` is very uncommon (in our app anyway):

```
[aaron@TC ~/g/github (master)]$ git grep '== nil' | wc -l
       5
[aaron@TC ~/g/github (master)]$ git grep 'nil ==' | wc -l
       0
[aaron@TC ~/g/github (master)]$ git grep '.nil?' | wc -l
    3135
```

**Comment by tenderlove** [comment]
Here's a breakdown of the instructions used at boot for our application:

```
> insn_info[order(insn_info$Count),]
              Instruction     Count
94                reverse         1
95       opt_newarray_max        49
93          setblockparam        54
92    opt_call_c_function        60
22          opt_aset_with       302
78              setglobal       417
39       setclassvariable       441
87                opt_div       597
88                   once      1263
91              branchnil      1357
52       opt_regexpmatch1      1386
53                duphash      1513
76              getglobal      1538
58                   topn      1607
56                opt_and      2129
89                 opt_or      2344
86               opt_mult      2859
46               toregexp     11460
28          definesmethod     11880
20            setconstant     12930
81                 intern     19087
5             defineclass     32953
55                opt_mod     33719
24          opt_aref_with     36994
80                  throw     44954
35               duparray     53448
96               setlocal     58770
30             getspecial     63765
69             opt_length     74235
77               newrange     77459
71               opt_size     98244
15           definemethod    104818
42               tostring    124320
70                 opt_gt    130296
74            concatarray    133235
48       getclassvariable    143407
34         opt_str_freeze    192432
68       opt_regexpmatch2    272036
90           checkkeyword    279475
44           freezestring    290999
14       putspecialobject    338706
43          concatstrings    344539
18              putstring    435151
45          getblockparam    455764
64            expandarray    490262
41              checktype    541355
21                newhash    698166
63                opt_neq    764259
57                   swap   1146326
66                 opt_ge   1283820
73                 opt_le   1345993
51              opt_nil_p   1536827
72              opt_minus   1598463
67            invokeblock   1614973
79     getblockparamproxy   1675137
82               getlocal   1799493
38               opt_ltlt   1888635
65             splatarray   1971596
83      opt_case_dispatch   2049897
85          setlocal_WC_1   2062346
50            invokesuper   2153631
84                 opt_lt   2229708
3      opt_setinlinecache   2422345
54               opt_plus   2803688
60            opt_empty_p   2820474
27               opt_aset   3204087
47   putobject_INT2FIX_0_   3790567
75            adjuststack   3924393
31                opt_not   3986585
2             getconstant   4722342
62                   jump   4781094
49                   dupn   4854287
37    setinstancevariable   5385254
61             checkmatch   5448507
36               newarray   5539621
26                   setn   6319773
59   putobject_INT2FIX_1_   9551850
16                defined  10071630
32          getlocal_WC_1  15007475
29                 opt_eq  16058544
10                   send  16372975
9               putobject  20392124
4                  putnil  26095091
19               branchif  30425098
33               opt_aref  31976718
1      opt_getinlinecache  32302934
40    getinstancevariable  34616623
13                    pop  36429507
7           setlocal_WC_0  39844133
23                    dup  41106221
11                    nop  51714288
8                 putself  56476479
17           branchunless  62428811
12                  leave  86843816
6  opt_send_without_block 147141336
25          getlocal_WC_0 213198244
```

This instruction is used about 1.5 million times during boot:

```
> insn_info[which(insn_info$Instruction == "opt_nil_p"),]
   Instruction   Count
51   opt_nil_p 1536827
```

**Comment by k0kubun** [comment]
That is interesting metrics and gives me some insights about which instructions to optimize. Thanks to share.

As you measured the boot of your application, could you also check its boot time before and after this patch? I'd guess at least it does not take so long compared to a measurement enabling `VM_COLLECT_USAGE_DETAILS`. If what or how to measure is not so straight-forward, it's okay.

**Comment by tenderlove** [comment]
Sorry it took me a while to get the data.  Our application takes nearly 60 seconds to boot up.  I compared ruby/ruby@369c36ef15c15ddc297e216522cc21599ed6a8e1 and this branch.  I timed the boot of our application 50 times on each branch and plotted the results:

plot

This patch saves about 1.8 seconds on our boot time.  I posted CSV files [here](https://gist.github.com/tenderlove/ebcd12dcd70048783f4395f4e3294718) if you want to check the numbers.

**Comment by tenderlove** [comment]
With Y at 0

plot

**Comment by ioquatix** [comment]
What was the outcome?

**Comment by ko1** [comment]
Boot time is interesting measurement. However, runtime (request) is more interest. Can you measure?

**Comment by tenderlove** [comment]
> Boot time is interesting measurement. However, runtime (request) is more interest. Can you measure?

I'll try, but it's not an easy number to get (right now 😞)

**Pull request closed without merging**